### DISCUSSION TITLE
IP6 support

### DISCUSSION THREAD

**Discussion by semuadmin** [semuconsulting/pygnssutils#27] [Ideas and Feature Requests]
(from Jan)

It would be useful if "gnssserver" and "pygnssutils - rtk_example.py" could be connected over IPv6.

**Comment by semuadmin** [comment]
Hi Jan,

IPv6 support in `gnssntripclient` (_as used in rtk_example.py_) is fairly easy to implement. I'll include this in the next release v1.0.8.

IPv6 support in `gnsssserver` is a little more difficult to test as it requires IPv6 on the host platform's network, which very few domestic LANs do at the moment.

If you want to beta test a preview of the next release of gnssntripclient, you can install pygnssutils from the [RC-1.0.8 branch](https://github.com/semuconsulting/pygnssutils/tree/RC-1.0.8):

```
pip install git+https://github.com/semuconsulting/pygnssutils.git@RC-1.0.8
```

**Reply by japs386** [reply]
The gnssserver RC-1.0.8 works well with

--inetmode IPv6 --hostip ::

With just

--inetmode IPv6

it fails with error message:

Starting output thread, broadcasting on 0.0.0.0:2101...
Error starting socket server [Errno -9] Address family for hostname not supported

Run on Rocky Linux 5.14.0-162.23.1.el9_1.x86_64

**Reply by semuadmin** [reply]
My bad - the default hostip is always 0.0.0.0 regardless of the inetmode setting. On it. Thanks for testing.

**Reply by semuadmin** [reply]
> The gnssserver RC-1.0.8 works well with
> 
> --inetmode IPv6 --hostip ::
> 
> With just
> 
> --inetmode IPv6
> 
> it fails with error message:
> 
> Starting output thread, broadcasting on 0.0.0.0:2101... Error starting socket server [Errno -9] Address family for hostname not supported
> 

Should be sorted now. NB: have changed --inetmode to --ipprot to avoid confusion with UBX message mode.

gnssserver -h will still list "0.0.0.0" as default host, but will convert to "::" if ipprot = IPv6. Also added --ipprot keyword to gnssdump (IPv6 address must be provided in the form [address]:port):

server (gnssserver):

```
➜  ~ gnssserver -I /dev/tty.usbmodem2101 --ipprot IPv6
2023-05-06 21:12:29.929915: Starting server (type CTRL-C to stop)...
2023-05-06 21:12:29.929941: Starting input thread, reading from /dev/tty.usbmodem2101...
2023-05-06 21:12:29.947395: Parsing GNSS data stream from: Serial<id=0x1054e33d0, open=True>(port='/dev/tty.usbmodem2101', baudrate=9600, bytesize=8, parity='N', stopbits=1, timeout=3, xonxoff=False, rtscts=False, dsrdtr=False)...

2023-05-06 21:12:30.435425: Starting output thread, broadcasting on :::50010...
2023-05-06 21:22:01.201543: Client ('::1', 54376, 0, 0) has connected. Total clients: 1
2023-05-06 21:22:07.040849: Client ('::1', 54376, 0, 0) has disconnected. Total clients: 0
```

client (gnssdump):

```
➜  ~ gnssdump -S "[::]:50010" --ipprot IPv6
2023-05-06 21:22:02.038765: Parsing GNSS data stream from: <socket.socket fd=3, family=30, type=1, proto=0, laddr=('::1', 54376, 0, 0), raddr=('::1', 50010, 0, 0)>...

<NMEA(GNRMC, time=20:22:02, status=A, lat=51.1206553333, NS=N, lon=-2.143435, EW=W, spd=0.012, cog=, date=2023-05-06, mv=, mvEW=, posMode=A, navStatus=V)>
...
```

**Comment by japs386** [comment]
The gnssserver RC-1.0.8 works well with or without "--hostip ::"
when "--ipprot IPv6" is set.

Notice that "--ipprot IPv6" also supports IPv4 connections:

Starting output thread, broadcasting on :::2101...
Client ('::ffff:77.241.136.161', 53452, 0, 0) has connected. Total clients: 1
Client ('::ffff:77.241.136.161', 53452, 0, 0) has disconnected. Total clients: 0

Run on Rocky Linux 5.14.0-162.23.1.el9_1.x86_64

**Comment by semuadmin** [comment]
Functionality now available in v1.0.8
